### PR TITLE
Add lsstSim PA1 test and serious bugfix tickets/DM-8551

### PR DESCRIPTION
**Pull request by parejkoj**
While writing the PA1 test for lsstSim, I discovered a bug that changed the results depending on whether it was run via unittest or from plot_jointcal_results.py. The photometric part of that is fixed (the astrometric part is a "feature" of data_ref ordering which is now noted, but probably not worth fixing).

### CODE REVIEW

File: python/lsst/jointcal/utils.py [hunk 1]
```diff
@@ -361,12 +365,12 @@ def plot_flux_distributions(plt, old_mag, new_mag, old_jitter, new_jitter,
     plt.savefig(filename.format(name))
 
 
-def plot_all_wcs_deltas(plt, data_refs, visit_catalogs, old_wcs_list, name,
+def plot_all_wcs_deltas(plt, data_refs, visits, old_wcs_list, name,
```
Comment (isullivan) [inline python/lsst/jointcal/utils.py]: Here and in the functions below, please define the parameters and return values.

File: tests/jointcalTestBase.py [hunk 2]
```diff
@@ -74,12 +75,15 @@ def _prep_reference_loader(self, center, radius):
         # Make a copy of the reference catalog for in-memory contiguity.
         self.reference = refLoader.loadSkyCircle(center, radius, filterName='r').refCat.copy()
 
-    def _testJointCalTask(self, nCatalogs, relative_error, absolute_error):
-        """Test parseAndRun for jointcal on nCatalogs, requiring less than some error (arcsec)."""
+    def _testJointCalTask(self, nCatalogs, dist_rms_relative, dist_rms_absolute, pa1):
```
Comment (isullivan) [inline tests/jointcalTestBase.py]: Please define the parameters and return value.

File: tests/test_jointcal_lsstSim.py [hunk 3]
```diff
@@ -54,32 +54,36 @@ def setUp(self):
...
     @unittest.skipIf(data_dir is None, "testdata_jointcal not setup")
     def testJointCalTask_2_visits(self):
         # NOTE: The relative RMS limits were empirically determined from the
         # first run of jointcal on this data. We should always do better than
         # this in the future!
-        relative_error = 9.7e-3*u.arcsecond
-        self._testJointCalTask(2, relative_error, absolute_error)
+        dist_rms_relative = 9.7e-3*u.arcsecond
+        pa1 = 2.64e-3
+        self._testJointCalTask(2, dist_rms_relative, dist_rms_absolute, pa1)
```
Comment (parejkoj) [inline tests/test_jointcal_lsstSim.py]: In general, I agree, but I think the intention of these magic numbers is pretty clear from the context of the method name.

Comment (isullivan) [inline tests/test_jointcal_lsstSim.py]: Mostly true, though at least in the test above (testJointCalTask_1_visits), I really don't know how those numbers are supposed to do what the test name says.

File: python/lsst/jointcal/utils.py [hunk 4]
```diff
@@ -51,39 +51,47 @@ def __init__(self, match_radius=0.1*arcseconds, flux_limit=100.0, verbose=False)
...
-            visit_catalogs = self._make_visit_catalogs(catalogs, visits_per_dataRef, visit_list)
-            catalogs = list(visit_catalogs.values())
+            visit_catalogs = self._make_visit_catalogs(catalogs, self.visits_per_dataRef)
+            catalogs = [visit_catalogs[x] for x in self.visits_per_dataRef]
             # use the first catalog as the relative reference catalog
+            # NOTE: The "first" catalog depends on the original ordering of the data_refs.
+            # NOTE: Thus, because I'm doing a many-1 match in _make_match_dict,
+            # the number of matches (and thus the details of the match statistics)
+            # will change if the data_refs are ordered differently.
+            # All the more reason to use a proper n-way matcher here...
```
Comment (parejkoj) [inline python/lsst/jointcal/utils.py]: I only have the epic for reworking it (which I've added a reference to): I don't see any tickets about actually writing an n-way matcher.

File: tests/test_jointcal_lsstSim.py (hunk 3 above)
Comment (parejkoj) [inline tests/test_jointcal_lsstSim.py]: Because that was a bug! Thanks for catching it.

Comment (parejkoj) [inline tests/test_jointcal_lsstSim.py]: (although its irrelevant, given the fact that I can't run jointcal on a single catalog at present anyway.)

File: python/lsst/jointcal/utils.py [hunk 5]
```diff
@@ -124,9 +130,10 @@ def rms_total(data):
         self.old_dist_total = MatchDict(*(tuple(map(rms_total, self.old_dist))*u.radian).to(u.arcsecond))
         self.new_dist_total = MatchDict(*(tuple(map(rms_total, self.new_dist))*u.radian).to(u.arcsecond))
 
-        return self.new_dist_total.relative, self.new_dist_total.absolute
+        Rms_result = collections.namedtuple("rms_result", ["dist_relative", "dist_absolute", "pa1"])
```
Comment (parejkoj) [inline python/lsst/jointcal/utils.py]: A namedtuple is like a "mini-class", so should be capitalized (see the [example in the docs](https://docs.python.org/2/library/collections.html#collections.namedtuple)). That said, I should have called it "Rms_result" in the namedtuple constructor, which I have fixed.

File: python/lsst/jointcal/utils.py [hunk 6]
```diff
@@ -183,7 +188,7 @@ def rms_per_source(data):
                            self.old_dist_total.relative, self.old_dist_total.absolute,
                            self.new_dist_total.relative, self.new_dist_total.absolute, name, outdir=outdir)
 
-        plot_all_wcs_deltas(plt, data_refs, visit_catalogs, old_wcs_list, name,
+        plot_all_wcs_deltas(plt, data_refs, self.visits_per_dataRef, old_wcs_list, name,
                             outdir=outdir, per_ccd_plot=per_ccd_plot)
 
         if interactive:
```
Comment (parejkoj) [inline python/lsst/jointcal/utils.py]: Yes: the goal is for interactive mucking with the plots and poking at the data, which only happens if the interactive flag is set (defaults to false).

File: tests/jointcalTestBase.py (hunk 2 above)
Comment (parejkoj) [inline tests/jointcalTestBase.py]: Thanks: this one definitely needed a better docstring (and this gave me a reason to search for, and find, the sublimeText AutoDocstring plugin!).

File: python/lsst/jointcal/utils.py (hunk 1 above)
Comment (parejkoj) [inline python/lsst/jointcal/utils.py]: Done. AutoDocstring is the best.

**Comment by parejkoj** [comment]
I've updated based on your comments. Please see if the new docstrings in the latest commit are ok (some of them feel redundant with the arg names, but I wasn't sure what else to say).

**Comment by isullivan** [comment]
The new docstrings add a lot of verbosity, but look helpful for the
hypothetical frustrated developer having to dig into jointcal to find out
why something is breaking.

**Pull request merged**